tests/smoke_tests/checkers.py

```diff
@@ -0,0 +1,206 @@
+    PREDICTION = "prediction"
+    PERSONAL = "personal"
+    GLOBAL = "global"
+    LOCAL = "local"
+
+
+class LossType(Enum):
+    LOSS = "loss"
+    CHECKPOINT = "checkpoint"
+    BACKWARD = "backward"
+    PROXIMAL = "proximal_loss"
+    GLOBAL = "global"
+    LOCAL = "local"
+
+
+class MetricChecker(ABC):
+    def assert_metrics_exist(self, logs: str) -> None:
+        log_lines = logs.split("\n")
+        for log_line in log_lines:
+            if self.should_check(log_line):
+                self.check(log_line)
+
+        assert self.found_all_metrics(), (
+            f"Full output:\n{logs}\n" f"[ASSERT ERROR] Metrics {self.to_dict()} not found in logs."
+        )
+
+    @abstractmethod
+    def should_check(self, log_line: str) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def check(self, log_line: str) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def found_all_metrics(self) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def to_dict(self) -> Dict[str, Union[str, float]]:
+        raise NotImplementedError
+
+
+class LossChecker(MetricChecker):
+    def __init__(
+        self,
+        loss: float,
+        loss_type: LossType = LossType.LOSS,
+        metric_type: Optional[MetricType] = None,
+        tolerance: float = DEFAULT_TOLERANCE,
+    ):
+        self.loss = loss
+        self.loss_type = loss_type
+        self.metric_type = metric_type
+        self.tolerance = tolerance
+        self.found_loss = False
+
+    def found_all_metrics(self) -> bool:
+        return self.found_loss
+
+    def should_check(self, log_line: str) -> bool:
+        if self.loss_type == LossType.LOSS:
+            return "losses_distributed" in log_line
+
+        # all other loss types:
+        assert self.metric_type is not None
+        return f"Client {self.metric_type.value} Losses" in log_line
+
+    def check(self, log_line: str) -> None:
+        losses = self.parse_losses(log_line)
+        has_loss = any([approx(loss, abs=self.tolerance) == self.loss for loss in losses])
+
+        if has_loss:
+            self.found_loss = True
+
+    def parse_losses(self, log_line: str) -> List[float]:
+        try:
+            if self.loss_type == LossType.LOSS:
+                keyword = "losses_distributed"
+                if keyword in log_line:
+                    losses_string = log_line.split(keyword)[-1]
+                    losses_list = eval(losses_string)
+                    return [loss_tuple[1] for loss_tuple in losses_list]
+            else:
+                keyword = f"{self.loss_type.value}:"
+                if keyword in log_line:
+                    after_keyword = log_line.split(keyword)[-1]
+                    loss_string = after_keyword[: after_keyword.find("\t")]
+                    return [eval(loss_string)]
+        except Exception:
+            # if any parsing error occurs, return empty
+            return []
+        return []
```

Comment:
Is this final return necessary? It might be, but I would have thought the try and except completed all possible return routes.